include/maliput_sparse/builder/builder.h

```diff
@@ -0,0 +1,219 @@
+ *         .StartLaneGeometry()
+ *           .LeftLineString(left_line_string_1)
+ *           .RightLineString(right_line_string_1)
+ *         .EndLaneGeometry()
+ *       .EndLane()
+ *       .StartLane()
+ *         .Id("j0_s0_l2")
+ *         .StartLaneGeometry()
+ *           .LeftLineString(left_line_string_2)
+ *           .RightLineString(right_line_string_2)
+ *         .EndLaneGeometry()
+ *       .EndLane()
+ *     .EndSegment()
+ *   .EndJunction()
+ *   .Build();
+ * @endcode{cpp}
+ */
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <maliput/api/junction.h>
+#include <maliput/api/lane.h>
+#include <maliput/api/road_geometry.h>
+#include <maliput/api/segment.h>
+#include <maliput/common/passkey.h>
+#include <maliput/geometry_base/junction.h>
+#include <maliput/geometry_base/lane.h>
+#include <maliput/geometry_base/road_geometry.h>
+#include <maliput/geometry_base/segment.h>
+
+#include "maliput_sparse/geometry/line_string.h"
+
+namespace maliput_sparse {
+namespace builder {
+namespace details {
+
+class BuilderBase {
+ public:
+  virtual ~BuilderBase() = default;
+};
+
+template <typename ParentT>
+class NestedBuilder : public BuilderBase {
+ public:
+  explicit NestedBuilder(ParentT* parent) : parent_(parent) { MALIPUT_THROW_UNLESS(parent_ != nullptr); }
+
+  ParentT& End() { return *parent_; }
+
+  ParentT* Parent() { return parent_; }
+
+ private:
+  ParentT* parent_{};
+};
+
+}  // namespace details
+
+class RoadGeometryBuilder;
+class JunctionBuilder;
+class SegmentBuilder;
+class LaneBuilder;
+
+/**
+ * @brief Builder class for LaneGeometry.
+ */
+class LaneGeometryBuilder final : public details::NestedBuilder<LaneBuilder> {
+ public:
+  explicit LaneGeometryBuilder(LaneBuilder* parent) : details::NestedBuilder<LaneBuilder>(parent) {}
+
+  // TODO(): implement me!
+  // LaneGeometryBuilder& LeftLineString(LineString3d&& left_line_string);
+  // LaneGeometryBuilder& RightLineString(LineString3d&& right_line_string);
+
+  LaneBuilder& EndLaneGeometry() { return End(); }
+};
+
+/**
+ * @brief Builder class for maliput::api::Lanes.
+ */
+class LaneBuilder final : public details::NestedBuilder<SegmentBuilder> {
+ public:
+  explicit LaneBuilder(SegmentBuilder* parent) : details::NestedBuilder<SegmentBuilder>(parent) {}
+
+  LaneBuilder& Id(const maliput::api::LaneId& lane_id);
+  LaneGeometryBuilder StartLaneGeometry();
+  SegmentBuilder& EndLane();
+
+  // TODO(): implement me!
+  // void SetLaneGeometry(maliput::common::Passkey<LaneGeometryBuilder>, std::unique_ptr<LaneGeometry> lane_geometry);
+ private:
+  maliput::api::LaneId id_{"l_id"};
+};
+
+/**
+ * @brief Builder class for maliput::api::Segments.
+ */
+class SegmentBuilder final : public details::NestedBuilder<JunctionBuilder> {
+ public:
+  explicit SegmentBuilder(JunctionBuilder* parent) : details::NestedBuilder<JunctionBuilder>(parent) {}
+
+  SegmentBuilder& Id(const maliput::api::SegmentId& segment_id);
+  LaneBuilder StartLane();
```

Comment:
How is it expected the lanes to be added? Related to the ordering of the lanes. Should the builder be smart enough in order to compute the ordering accordingly to the lane strings?

Reply:
Good question.

I think the 2 target maps we have in mind now have a clear sense of left to right ordering, so we could let the parser derive that who has more understanding and context instead of the builder which needs to rely on lane specific information.

Alternatively, we could change this behavior by adding a sorting policy to the SegmentBuilder that is set before calling EndSegment(). That will make the EndSegment() function to sort lanes in a specific way.

Reply:
> I think the 2 target maps we have in mind now have a clear sense of left to right ordering, so we could let the parser derive that who has more understanding and context instead of the builder which needs to rely on lane specific information.

+1